fl4health/servers/base_server.py

```diff
@@ -84,9 +148,103 @@ def report_centralized_eval(self, history: History, num_rounds: int) -> None:
+        if self.per_round_checkpointer.checkpoint_exists():
+            self._load_server_state()
+        else:
+            log(INFO, "Initializing server state")
+            self.parameters = self._get_initial_parameters(server_round=0, timeout=timeout)
+            self.history = History()
+            self.current_round = 1
+
+        if self.current_round == 1:
+            log(INFO, "Evaluating initial parameters")
+            res = self.strategy.evaluate(0, parameters=self.parameters)
+            if res is not None:
+                log(
+                    INFO,
+                    "initial parameters (loss, other metrics): %s, %s",
+                    res[0],
+                    res[1],
+                )
+                self.history.add_loss_centralized(server_round=0, loss=res[0])
+                self.history.add_metrics_centralized(server_round=0, metrics=res[1])
+
+            # Run federated learning for num_rounds
+            log(INFO, "FL starting")
+
+        start_time = datetime.datetime.now()
+
+        while self.current_round < (num_rounds + 1):
+            # Train model and replace previous global model
+            res_fit = self.fit_round(server_round=self.current_round, timeout=timeout)
+            if res_fit:
+                parameters_prime, fit_metrics, _ = res_fit  # fit_metrics_aggregated
+                if parameters_prime:
+                    self.parameters = parameters_prime
+                self.history.add_metrics_distributed_fit(server_round=self.current_round, metrics=fit_metrics)
+
+            # Evaluate model using strategy implementation
+            res_cen = self.strategy.evaluate(self.current_round, parameters=self.parameters)
+            if res_cen is not None:
+                loss_cen, metrics_cen = res_cen
+                log(
+                    INFO,
+                    "fit progress: (%s, %s, %s, %s)",
+                    self.current_round,
+                    loss_cen,
+                    metrics_cen,
+                    (datetime.datetime.now() - start_time).total_seconds(),
+                )
+                self.history.add_loss_centralized(server_round=self.current_round, loss=loss_cen)
+                self.history.add_metrics_centralized(server_round=self.current_round, metrics=metrics_cen)
+
+            # Evaluate model on a sample of available clients
+            res_fed = self.evaluate_round(server_round=self.current_round, timeout=timeout)
+            if res_fed:
+                loss_fed, evaluate_metrics_fed, _ = res_fed
+                if loss_fed:
+                    self.history.add_loss_distributed(server_round=self.current_round, loss=loss_fed)
+                    self.history.add_metrics_distributed(server_round=self.current_round, metrics=evaluate_metrics_fed)
+
+            self.current_round += 1
+
+            # Save checkpoint after training and testing
+            self._hydrate_model_for_checkpointing()
```

Comment:
It's a good question. It's because of the state save part actually. Because I reordered the methods (shouldn't have done that for the diffing 😂), it's a bit hidden, but we had this method previously as well. Basically, we need to intervene to maybe load state at the beginning of a round, save state at the end of the rounds and to keep track of what round we're on etc.

I don't think I'm actually changing anything in this method. I just moved it and renamed the state saving functions to have the _ prefix to indicate an internal call.

Reply:
Ah yes i'm reminded again why this is necessary. Annoying that the whole method needs to overridden to just so that a single line can be changed to resume from a particular round.